Finding best model and hyper parameter tunning using GridSearchCV for the digit dataset in sklearn.dataset

In [1]:
from sklearn.datasets import load_digits
from sklearn import svm
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.naive_bayes import GaussianNB, MultinomialNB
from sklearn.tree import DecisionTreeClassifier
import pandas as pd
from sklearn.model_selection import GridSearchCV  

In [3]:
digits = load_digits()

In [10]:
digits_df = pd.DataFrame(digits.data, columns=[f'pixel_{i}' for i in range(digits.data.shape[1])])
digits_df['target'] = digits.target
print(digits_df.head())

   pixel_0  pixel_1  pixel_2  pixel_3  pixel_4  pixel_5  pixel_6  pixel_7  \
0      0.0      0.0      5.0     13.0      9.0      1.0      0.0      0.0   
1      0.0      0.0      0.0     12.0     13.0      5.0      0.0      0.0   
2      0.0      0.0      0.0      4.0     15.0     12.0      0.0      0.0   
3      0.0      0.0      7.0     15.0     13.0      1.0      0.0      0.0   
4      0.0      0.0      0.0      1.0     11.0      0.0      0.0      0.0   

   pixel_8  pixel_9  ...  pixel_55  pixel_56  pixel_57  pixel_58  pixel_59  \
0      0.0      0.0  ...       0.0       0.0       0.0       6.0      13.0   
1      0.0      0.0  ...       0.0       0.0       0.0       0.0      11.0   
2      0.0      0.0  ...       0.0       0.0       0.0       0.0       3.0   
3      0.0      8.0  ...       0.0       0.0       0.0       7.0      13.0   
4      0.0      0.0  ...       0.0       0.0       0.0       0.0       2.0   

   pixel_60  pixel_61  pixel_62  pixel_63  target  
0      10.0     

In [13]:
X = digits_df.drop('target', axis=1)  
y = digits_df['target']  

In [17]:
scores = []

for model_name, mp in model_params.items():
    clf =  GridSearchCV(mp['model'], mp['params'], cv=5, return_train_score=False)
    clf.fit(X,y)
    scores.append({
        'model': model_name,
        'best_score': clf.best_score_,
        'best_params': clf.best_params_
    })
    
test_df1 = pd.DataFrame(scores,columns=['model','best_score','best_params'])
test_df1

model  best_score                                best_params
0                  svm    0.947697               {'C': 1, 'kernel': 'linear'}
1        random_forest    0.897631                       {'n_estimators': 10}
2  logistic_regression    0.922114                                   {'C': 1}
3          gaussian_nb    0.806928                                         {}
4       multinomial_nb    0.870350                                         {}
5        decision_tree    0.815266  {'criterion': 'entropy', 'max_depth': 10}

In [30]:
model_params = {
    'svm': {
        'model': svm.SVC(),
        'params': {
            'C': [0.1, 1, 10],
            'kernel': ['linear', 'rbf'],
            'gamma': [1,2,3]
        }
    },
    'random_forest': {
        'model': RandomForestClassifier(),
        'params': {
            'n_estimators': [50, 100, 150],
            'max_depth': [None, 10, 20, 30]
        }
    },
    'logistic_regression': {
        'model': LogisticRegression(),
        'params': {
            'C': [0.1, 1, 10],
            'solver': ['liblinear', 'lbfgs'],
            'max_iter': [100, 200, 300, 500]
        }
}


In [31]:
scores = []

for model_name, mp in model_params.items():
    clf =  GridSearchCV(mp['model'], mp['params'], cv=5, return_train_score=False)
    clf.fit(X,y)
    scores.append({
        'model': model_name,
        'best_score': clf.best_score_,
        'best_params': clf.best_params_
    })
    
test_df1 = pd.DataFrame(scores,columns=['model','best_score','best_params'])
test_df1

model  best_score                                best_params
0   svm    0.972187  {'C': 1, 'gamma': 0.001, 'kernel': 'rbf'}

So the best model here is SVM( C = 1 , gamma = 0.001, kernel = rbf)
 with 97.2% accuracy 